index.js

```diff
@@ -199,6 +200,7 @@ Router.prototype.handle = function handle(req, res, callback) {
       req.baseUrl = parentUrl
       req.url = protohost + removed + req.url.substr(protohost.length)
       removed = ''
+      req.matchedRoutes = parentMatchedRoutes
```

Comment:
It seems like this is trying to restore the `req.matchedRoutes` back to what it used to be, but since the router is just directly manipulating the array, it still ends up containing the new additions that were made. I'm not 100% sure what the purpose of this restoration of the property is meant to do after every route call?

Reply:
@dougwilson This whole approach needs to be tweaked anyway to to deal with the "sibling" routes example you mentioned in https://github.com/pillarjs/router/pull/86#issuecomment-570938956. 

I'm open to suggestions for the best way to deal with the `matchedRoutes` "stack". I don't know enough yet about the router internals to know the best place to push and pop from it.

Reply:
Well, due to the lack of documentation here or a full written explanation for how this all should work, I really have no idea what to suggest for the changes, because I'm still unclear how this feature is actually supposed to operate :) Maybe we should circle back to explaining exactly how this feature should work in text form, so then we can all help on the implementation by knowing what we're trying to achieve. Right now, I'm still unclear on what exactly the plain is, so don't have a way to suggest what changes to make to the code to get the outcome to be whatever it is it should be, if that makes sense.

Reply:
@dougwilson Apologies. This PR has moved a number of times between expressjs and pillarjs, so much of the original context may have been lost.

AFAIK, this all got started with https://github.com/expressjs/express/issues/2501 requesting `req.originalRoute` to complement `req.originalUrl`. The use case provided there is for logging route names instead of fully resolved paths, which is the same as what I need. Another use case I've seen is for recording metrics. E.g. when recording how many times an endpoint like `GET /users/:id` is called, you may want to use the full name of the endpoint without path parameters substituted. 

As for the technical decisions in this PR, I can't speak to most of them. The original author (@ajfranzoia) seems to have abandoned his efforts, and I'm merely picking what he wrote back up. Perhaps @wesleytodd can provide more insight?

I assume `req.originalRoute` morphed into an array of `matchedRoutes` when it was realized that route names aren't always strings and that it wasn't as simple as joining path segments together.